semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py

```diff
@@ -131,31 +121,12 @@ def parse(yaml_file):
             prefix = group.get("prefix", "")
             if prefix != "":
                 validate_id(prefix, group.lc.data["prefix"])
-            position = group.lc.data["id"]
-            # Fields validation based on the type
-            if convention_type == SemanticConventionType.RESOURCE:
-                if span_kind != SpanKind.EMPTY:
-                    position = group.lc.data["span_kind"]
-                    msg = "Resources cannot have span_kind: {}".format(
-                        group.get("span_kind")
-                    )
-                    raise ValidationError.from_yaml_pos(position, msg)
-            model = SemanticConvention(
-                semconv_id=group["id"].strip(),
-                type=convention_type,
-                brief=str(group["brief"]).strip(),
-                note=group.get("note", "").strip(),
-                prefix=prefix.strip(),
-                extends=group.get("extends", "").strip(),
-                span_kind=span_kind,
-                attrs_by_name=SemanticAttribute.parse(prefix, group.get("attributes"))
-                if "attributes" in group
-                else {},
-                constraints=SemanticConvention.parse_constraint(
-                    group.get("constraints", ())
-                ),
-                _position=position,
-            )
+
+            # First, validate that the correct fields are available in the yaml
+            convention_type.validate_keys(group)
+            model = convention_type(group)
```

Comment:
Should the convention class validate itself in the constructor so you don't have to remember to validate separately?

Reply:
I considered this. The logical place to put this would be in `ValidatableYamlNode.__init__`, because that superclass owns the validation logic. **..._but_**, the subclasses are likely adding attributes in their own init methods, so the correct place to call it would be _after_ the superclass and subclass init methods had completed.

Reply:
maybe in a factory then?

Reply:
I moved the logic to a factory function.
This triggered some renaming (so I wouldn't have a naming collision between `def SemanticConvention` and `class SemanticConvention`), which had some cascading effects.  
I removed several `import SemanticConvention` lines where we were only using the imported class for type declarations, and I refactored references to `SemanticConvention.parse` to a new `parse_semantic_convention_groups` function (and in the process I split the parsing concern away from the data struct concern).